bindings/rust/wasmedge-sys/src/module.rs

```diff
@@ -1,41 +1,45 @@
 use super::wasmedge;
-use crate::raw_result::{decode_result, ErrReport};
+use crate::{
+    error::{check, WasmEdgeResult},
+    utils, Config,
+};
 use ::core::mem::MaybeUninit as MU;
-use std::ffi::CString;
+use std::path::Path;
 
 #[derive(Debug)]
 pub struct Module {
     pub(crate) ctx: *mut wasmedge::WasmEdge_ASTModuleContext,
+    pub(crate) registered: bool,
 }
-
 impl Drop for Module {
     fn drop(&mut self) {
-        unsafe { wasmedge::WasmEdge_ASTModuleDelete(self.ctx) };
+        if !self.registered && !self.ctx.is_null() {
+            unsafe { wasmedge::WasmEdge_ASTModuleDelete(self.ctx) };
+        }
     }
 }
-
 impl Module {
-    pub fn load_from_file(
-        config: &crate::config::Config,
-        path: CString,
-    ) -> Result<Self, ErrReport> {
+    pub fn load_from_file<P: AsRef<Path>>(config: &Config, path: P) -> WasmEdgeResult<Self> {
         let loader_ctx = unsafe { wasmedge::WasmEdge_LoaderCreate(config.ctx) };
         let mut ctx: *mut wasmedge::WasmEdge_ASTModuleContext = std::ptr::null_mut();
 
-        let res = unsafe {
-            wasmedge::WasmEdge_LoaderParseFromFile(loader_ctx, &mut ctx as *mut _, path.as_ptr())
-        };
-        decode_result(res)?;
+        let path = utils::path_to_cstring(path.as_ref())?;
 
-        assert!(!ctx.is_null(), "WasmEdge failed to load from file!");
+        unsafe {
+            check(wasmedge::WasmEdge_LoaderParseFromFile(
+                loader_ctx,
+                &mut ctx as *mut _,
+                path.as_ptr(),
+            ))?;
+        }
 
-        Ok(Self { ctx })
+        Ok(Self {
+            ctx,
+            registered: false,
+        })
     }
 
-    pub fn load_from_buffer(
-        config: &crate::config::Config,
-        buffer: Vec<u8>,
-    ) -> Result<Self, ErrReport> {
+    pub fn load_from_buffer(config: &crate::config::Config, buffer: &[u8]) -> WasmEdgeResult<Self> {
         let loader_ctx = unsafe { wasmedge::WasmEdge_LoaderCreate(config.ctx) };
         let mut ctx: *mut wasmedge::WasmEdge_ASTModuleContext = std::ptr::null_mut();
 
```

Comment:
nice! This assertion is guaranteed to be safety. 
This unsafe function should be marked with a document comment to indicate that it is safe, as is the usual practice. 
```rust
/// # Safety 
///  Avoid the problem of uninitialized memory due to the introduction of an empty buffer
```